Source/DafnyPipeline.Test/Formatter.cs

```diff
@@ -0,0 +1,2611 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Diagnostics.Contracts;
+using System.IO;
+using System.Text.RegularExpressions;
+using JetBrains.Annotations;
+using Bpl = Microsoft.Boogie;
+using BplParser = Microsoft.Boogie.Parser;
+using Microsoft.Dafny;
+using Xunit;
+
+namespace DafnyPipeline.Test {
+
+  // Simple test cases (FormatterWorksFor with only one argument)
+  // consist of removing all the indentation from the program,
+  // adding it through the formatter, and checking if we obtain the initial result
+  //
+  // Advanced test cases consist of passing the program and its expected result after indentation
+  //
+  // Every test is performed with all three newline styles
+  // Every formatter program is formatted again to verify that it stays the same.
+  [Collection("Singleton Test Collection - Formatter")]
+  public class Formatter {
+    enum Newlines {
+      LF,
+      CR,
+      CRLF
+    };
+
+    private static Regex indentRegex = new Regex(@"(?<=\n|\r(?!\n))[ \t]*");
+
+    private static Regex removeTrailingNewlineRegex = new Regex(@"(?<=\S|\r|\n)[ \t]+(?=\r?\n|\r(?!\n)|$)");
+
+    private Newlines currentNewlines;
+
+    [Fact]
+    public void FormatterWorksForFinalSpaceAfterDatatype() {
+      FormatterWorksFor(@"
+datatype Maybe<T> = None | Some(get: T)
+");
+    }
+
+    [Fact]
+    public void FormatterWorksForDecreasesInvariantComments() {
+      FormatterWorksFor(@"
+method Test() {
+  while X
+    decreases true || false
+    // comment
+    invariant true || false
+    // comment
+  {
+  }
+}
+");
+    }
+
+
+
+    [Fact]
+    public void FormatterWorksForArguments() {
+      FormatterWorksFor(@"
+method Test()
+{
+  me(func(arg(3,
+              4)));
+
+  me(func(arg(5,
+              6
+             ,7
+          )
+     )
+  );
+
+  me(func(arg(
+            8, 9)));
+
+  me(func(
+       arg(
+         10, 11)));
+
+  me(
+    func(
+      arg(
+        12, 13)),
+    func2());
+  me
+  (func
+   (arg
+    (1
+    ,2)),
+   func2());
+}
```

Comment:
I'm concerned that, when the formatter grows sophisticated enough to care about e.g. comma placement and line breaks, we're going to have to change a _lot_ of this file's tests - and when we do so we might not remember which features of the existing formatting we're exercising in each test/statement/expression.

Reply:
I think this is a reasonable concern, but what we will surely do in this case is add new tests for the new formatter, and keep these tests with a previous compatibility mode.
I have heard some push back from actually adding or removing newlines as well, so it's not something we might want to do immediately anyway.